services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -12,15 +14,49 @@ import scala.concurrent.{ ExecutionContext, Future }
 class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     extends ProjectVersioning[VersioningException] {
 
-  override def updateFile(project: Project, projectFile: ProjectFile)(
+  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[Version])(
     implicit ec: ExecutionContext
-  ): AsyncResult[String] = ???
+  ): AsyncResult[String] = {
+    val path = URLEncoder.encode(projectFile.path.toString, "UTF-8")
+    val fileUrl = s"${config.url}projects/${project.repository}/repository/files/$path"
+    val fileVersion = version match {
+      case Some(version) => version.value
+      case None          => config.defaultFileVersion
+    }
+
+    httpClient
+      .put(
+        fileUrl,
+        payload = Json.stringify(Json.toJson(UpdateFileRequest(fileVersion, projectFile.content))),
+        headers = config.token
+      )
+      .flatMap {
+        case Response(HttpStatusCodes.OK, _, _) =>
+          Future.successful(Right("Success update file"))
+        case Response(HttpStatusCodes.BadRequest, _, _) =>
+          httpClient
+            .post(
+              fileUrl,
+              payload = Json.stringify(
+                Json.toJson(CreateFileRequest(config.defaultFileVersion, projectFile.content, "Init commit"))
+              ),
+              headers = config.token
```

Comment:
Please double check the behaviour here. 
Are you sure it's possible scenario? By default `PUT` http method updates file **or creates new one, if the file does not exist**.

Reply:
Will check it. GitLab api doc saying only about update behavior when file is exist